test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -2702,90 +2734,157 @@ class test_ucp_sockaddr_protocols_err_sender
+        std::string recv_copy(size, 'y');
+        ucs_status_t status;
+
+        using send_recv_info_t = std::tuple<void*, ucp_tag_message_h, void*>;
+        std::deque<send_recv_info_t> send_recv_infos;
+
+        for (size_t sender_idx = 0; sender_idx < num_senders; ++sender_idx) {
+            if (sender_idx > 0) {
+                client_ep_connect(sender_idx);
+                send_recv(sender(), sender_idx, receiver(), send_recv_type(),
+                          false, cb_type());
+            }
+
+            void *sreq = send(sender(), sender_idx, &send_buf[0], size,
+                              SEND_RECV_TAG, scomplete_cbx, NULL);
+            ASSERT_TRUE(UCS_PTR_IS_PTR(sreq));
+            ASSERT_EQ(UCS_INPROGRESS, ucp_request_check_status(sreq));
+
+            /* Allow receiver to get RTS notification, but do not receive message
+             * body */
+            ucp_tag_recv_info_t info;
+            ucp_tag_message_h message = message_wait(receiver(), 0, 0, &info);
+            ASSERT_NE((void*)NULL, message);
+            ASSERT_EQ(UCS_INPROGRESS, ucp_request_check_status(sreq));
+
+            send_recv_infos.emplace_back(sreq, message, nullptr);
+        }
+
+        /* Ignore all errors - it is expected */
+        scoped_log_handler slh(wrap_errors_logger);
+
+        /* Close the first sender's EP to force send operation to be completed
+         * with CANCEL status */
+        entity_disconnect(sender());
+
+        if (has_invalidate && (num_senders > 1)) {
+            for (auto &send_recv_info : send_recv_infos) {
+                /* Check that the send request would not be completed until the
+                 * memory handle for the same send buffer will not be released
+                 */
+                void *sreq          = std::get<0>(send_recv_info);
+                ucs_status_t status = request_progress(sreq, { &sender() },
+                                                       0.5);
+                ASSERT_EQ(UCS_INPROGRESS, status);
+            }
+        }
+
+        auto first_send_recv_info = std::move(send_recv_infos.front());
+        send_recv_infos.pop_front();
+
+        for (auto &send_recv_info : send_recv_infos) {
+            auto message                = std::get<1>(send_recv_info);
+            std::get<2>(send_recv_info) = recv(receiver(), &extra_recv_buf[0],
+                                               size, message,
+                                               rtag_complete_cbx, NULL);
+        }
+
+        while (!send_recv_infos.empty()) {
+            auto &send_recv_info = send_recv_infos.front();
+
+            status = requests_wait({ std::get<0>(send_recv_info),
+                                     std::get<2>(send_recv_info) });
+            ASSERT_UCS_OK(status);
+            send_recv_infos.pop_front();
+        }
+
+        void *sreq = std::get<0>(first_send_recv_info);
+        status     = request_wait(sreq);
+        ASSERT_EQ(UCS_ERR_CANCELED, status);
```

Comment:
maybe do this as part of the previous loops?
for example:
loop: go over send_recv_infos and post receive operations. add them to std::vector of requests
(the callback of sender0 can check if recv_buf==send_buf or status==OK. rest of callbacks should expect status==OK)
request_wait(all-requests: all sends and all receives)
check data

Reply:
done